src/search/parser/abstract_syntax_tree.h

```diff
@@ -0,0 +1,99 @@
+#ifndef PARSER_ABSTRACT_SYNTAX_TREE_H
+#define PARSER_ABSTRACT_SYNTAX_TREE_H
+
+#include "enriched_abstract_syntax_tree.h"
+#include "lexical_analyzer.h"
```

Comment:
I'm not sure what the state of the code was when you made this comment. We now have the following split:

* `lexical_analyzer.{h,cc}` generates a token stream defined in `token_stream.{h.cc}`
* `syntax_analyzer.{h,cc}` generates an AST defined in `abstract_syntax_tree.{h.cc}`
* `abstract_syntax_tree.{h.cc}` generates an enriched AST defined in `enriched_abstract_syntax_tree.{h.cc}`
* `enriched_abstract_syntax_tree.{h.cc}` constructs the actual objects as `plugins::Any` defined in `plugins/any.{hcc}`

So parsing code and the generated data structures are mostly separated except that the AST data structures of one level also contain the code to construct the next level. This is also because the implementation with virtual methods is easier than a visitor pattern. Again, I suggest we leave it like this for now.